src/utils/eval_tuner.ts

```diff
@@ -0,0 +1,224 @@
+import { getRandomItem } from '.'
+import { setRandomSeed } from 'bun:jsc'
+import type { Tensor } from '@shumai/shumai'
+
+setRandomSeed(Math.random() * 100000)
+
+export enum EvalTunerType {
+  STOCH_HILL_CLIMB
+}
+
+/* TODO: more algos supported & tinker w hillclimb defaults */
+export class EvalTuner {
+  public type: EvalTunerType
+  public best_time: number
+  public last_i: number
+  public done = false
+
+  private evalFn: StochHillClimb
+
+  private _last_time: number
+
+  set last_time(value: number) {
+    this._last_time = value
+  }
+
+  constructor(
+    type: EvalTunerType = EvalTunerType.STOCH_HILL_CLIMB,
+    lowerBound = 1,
+    upperBound = 500
+  ) {
+    this.type = type
+    switch (type) {
+      case EvalTunerType.STOCH_HILL_CLIMB:
+        this.evalFn = new StochHillClimb(lowerBound, upperBound)
+        break
+      default:
+        throw new Error('EvalTunerType: not supported')
+    }
+    this.best_time = Infinity
+    this.last_i = 0
+  }
+
+  get best_delta() {
+    return this.evalFn.bestIn
+  }
+
+  step(t: Tensor, i: number, usedDelta?: number) {
+    if (!usedDelta) usedDelta = this.evalFn.done ? this.evalFn.bestIn : this.evalFn.currentIn
+    const delta = i - this.last_i
+    if (delta >= usedDelta) {
+      // see how well we did this time
+      t.eval()
+      const time_spent = (performance.now() - this._last_time) / delta
+      // console.log('delta', usedDelta, 'took', time_spent, 'ms per increment')
+      if (this.evalFn.update(time_spent)) this.done = true
+      this.last_i = i
+      this._last_time = performance.now()
+    }
+  }
+}
+
+class StochHillClimb {
```

Comment:
worth putting a full algorithm explainer here :) quite a lot of params!

Reply:
Agreed; will do that and push w the FastOptimizer code this evening (will do a full algo explanation for FastOptimizer as well).

Reply:
Finished basic write up detailing the logic for this implementation; will be included in next commit.